nrf52-hal-common/src/spim.rs

```diff
@@ -270,105 +251,153 @@ impl<T> Spim<T> where T: SpimExt {
+        let rxi = rx_buffer.chunks_mut(EASY_DMA_SIZE);
+
+        txi.zip(rxi).try_for_each(|(t, r)| {
+            self.do_spi_dma_transfer(DmaSlice::from_slice(t), DmaSlice::from_slice(r), |cs| {
+                if cs {
+                    chip_select.set_low()
+                } else {
+                    chip_select.set_high()
+                }
+            })
+        })
+    }
 
-            // rx_len:
-            rx_buffer.len() as _,
-            // chip select callback
-            |cs|{if cs {chip_select.set_low()} else {chip_select.set_high()} }
-        )
+    /// Read and write from a SPI slave, using separate read and write buffers
+    ///
+    /// This method implements a complete read transaction, which consists of
+    /// the master transmitting what it wishes to read, and the slave responding
+    /// with the requested data.
+    ///
+    /// Uses the provided chip select pin to initiate the transaction. Transmits
+    /// all bytes in `tx_buffer`, then receives bytes until `rx_buffer` is full.
+    ///
+    /// This method is more complicated than the other `transfer` methods because
+    /// it is allowed to perform transactions where `tx_buffer.len() != rx_buffer.len()`.
+    /// If this occurs, extra incoming bytes will be discarded, OR extra outgoing bytes
+    /// will be filled with the `orc` value.
+    pub fn transfer_split_uneven(
+        &mut self,
+        chip_select: &mut Pin<Output<PushPull>>,
+        tx_buffer: &[u8],
+        rx_buffer: &mut [u8],
+    ) -> Result<(), Error> {
+        ram_slice_check(tx_buffer)?;
+        ram_slice_check(rx_buffer)?;
+        // For the tx and rx, we want to return Some(chunk)
+        // as long as there is data to send. We then chain a repeat to
+        // the end so once all chunks have been exhausted, we will keep
+        // getting Nones out of the iterators
+        let txi = tx_buffer
+            .chunks(EASY_DMA_SIZE)
+            .map(|c| Some(c))
+            .chain(repeat_with(|| None));
+
+        let rxi = rx_buffer
+            .chunks_mut(EASY_DMA_SIZE)
+            .map(|c| Some(c))
+            .chain(repeat_with(|| None));
+
+        // We then chain the iterators together, and once BOTH are feeding
+        // back Nones, then we are done sending and receiving
+        txi.zip(rxi)
+            .take_while(|(t, r)| t.is_some() && r.is_some())
+            // We also turn the slices into either a DmaSlice (if there was data), or a null
+            // DmaSlice (if there is no data)
+            .map(|(t, r)| {
+                (
+                    t.map(|t| DmaSlice::from_slice(t))
+                        .unwrap_or_else(|| DmaSlice::null()),
+                    r.map(|r| DmaSlice::from_slice(r))
+                        .unwrap_or_else(|| DmaSlice::null()),
+                )
+            })
+            .try_for_each(|(t, r)| {
+                self.do_spi_dma_transfer(t, r, |cs| {
+                    if cs {
```

Comment:
TODO: We probably don't want to toggle CS between each block, just once for the whole transaction